### PR TITLE
Extract Methods for Downloadable Classes

### PR DESCRIPTION
**Pull request by samuel-tate**
# Related Issues

#29 

## Testing Locally

```
$ mkdir foo
$ cd foo
$ touch bar
$ echo "bar bar bar bar" > bar
$ cd ..
$ zip -r foo foo
$ rm -r foo
$ python

>>> import pyrefinebio
>>> ds = pyrefinebio.Dataset()
>>> ds._downloaded_path = "./foo.zip"
>>> ds.extract()
```
you can see that `./foo.zip` was extracted into `./foo/` and that the bar file should also be there with the correct contents.

### CODE REVIEW

File: pyrefinebio/compendia.py [hunk 1]
```diff
@@ -95,6 +98,29 @@ def download(self, path, prompt=True):
 
             prompt (bool): if true, will prompt before downloading files bigger than 1GB
         """
-        self.computed_file.download(path, prompt)
+        if not self.computed_file.download_url:
+            raise DownloadError("Compendia", "Download url not found - did you set up and activate a Token?")
+
+        full_path = expand_path(path, "compendia-" + str(self.id) + ".zip")
```
Comment (kurtwheeler) [inline pyrefinebio/compendia.py:104]: Shouldn't we check to see if the path is a file or a directory before appending the filename to it?

File: pyrefinebio/compendia.py [hunk 2]
```diff
@@ -95,6 +98,29 @@ def download(self, path, prompt=True):
...
+            Compendium
+        """
+        if not self._downloaded_path:
+            raise Exception(
+                "No downloaded path exists. "
+                "Make sure you have successfully downloaded the Compendium before extracting."
+            )
+
+        extract(self._downloaded_path)
+        return self
```
Comment (kurtwheeler) [inline pyrefinebio/compendia.py:126]: Missing newline.

File: pyrefinebio/compendia.py [hunk 3]
```diff
@@ -95,6 +98,29 @@ def download(self, path, prompt=True):
...
+
+
+    def extract(self):
+        """Extract a downloaded Compendium
+
+        Returns:
+            Compendium
+        """
+        if not self._downloaded_path:
+            raise Exception(
```
Comment (kurtwheeler) [inline pyrefinebio/compendia.py:120]: Can we use a less generic exception class?

File: pyrefinebio/util.py [hunk 4]
```diff
@@ -14,6 +17,20 @@ def parse_date(date):
...
+    full_path = os.path.abspath(user_path)
+
+    if os.path.isdir(full_path):
+        full_path = os.path.join(full_path, filename)
+
+    return full_path
+
+
+def extract(path):
+    shutil.unpack_archive(path)
```
Comment (kurtwheeler) [inline pyrefinebio/util.py:31]: I think it'd be better if you didn't rename this function. Reading the other files, it's less clear what it's doing until you come here. Also calling an `extract` function from the `extract` methods is slightly confusing like "wait, is this recursive? oh no I'm in a class".

File: tests/test_compendia.py [hunk 5]
```diff
@@ -152,3 +152,14 @@ def test_compendium_download(self, mock_request):
...
+
+    @patch("pyrefinebio.util.shutil.unpack_archive")
+    def test_compendium_extract(self, mock_unpack):
+        c = pyrefinebio.Compendium()
+
+        c._downloaded_path = "foo"
+        
+        c.extract()
+
+        mock_unpack.assert_called_with("foo")
```
Comment (kurtwheeler) [inline tests/test_compendia.py:165]: Missing newline.

**Review by kurtwheeler — CHANGES_REQUESTED** [review]
I think you should add an `extract=False` parameter to higher level functions.

File: pyrefinebio/compendia.py (hunk 1 above)
Comment (samuel-tate) [reply pyrefinebio/compendia.py:104]: `expand_path` should [already do this](https://github.com/AlexsLemonade/refinebio-py/pull/39/files#diff-fbbdc042b88d5d41b2743515f83a29f5c1e8e794dbc321196453a87f2722143eR24-R25).

It'll only append the filename part if `path` is a dir.

Comment (kurtwheeler) [reply pyrefinebio/compendia.py:104]: Oh beautiful!

File: pyrefinebio/compendia.py (hunk 3 above)
Comment (samuel-tate) [reply pyrefinebio/compendia.py:120]: I wasn't sure which exception to use.
maybe `DownloadError` or should I make a new class for extraction related errors?

Comment (kurtwheeler) [reply pyrefinebio/compendia.py:120]: Mmm maybe a new class like MissingFile or something?

**Review by kurtwheeler — APPROVED** [review]
Sweet, this LGTM!

File: pyrefinebio/compendia.py (hunk 3 above)
Comment (samuel-tate) [reply pyrefinebio/compendia.py]: added MissingFile exception class

**Pull request merged**